include/rocksdb/table.h

```diff
@@ -51,6 +51,55 @@ enum ChecksumType : char {
   kxxHash64 = 0x3,
 };
 
+// `PinningTier` is used to specify which tier of block-based tables should
+// be affected by a block cache pinning setting (see
+// `MetadataCacheOptions` below).
+enum class PinningTier {
+  // For compatibility, this value specifies to fallback to the behavior
+  // indicated by the deprecated options,
+  // `pin_l0_filter_and_index_blocks_in_cache` and
+  // `pin_top_level_index_and_filter`.
+  kFallback,
+
+  // This tier contains no block-based tables.
+  kNone,
+
+  // This tier contains block-based tables that may have originated from a
+  // memtable flush. In particular, it includes tables from L0 that are smaller
+  // than 1.5 times the current `write_buffer_size`. Note these criteria imply
+  // it can include intra-L0 compaction outputs and ingested files, as long as
+  // they are not abnormally large compared to flushed files in L0.
+  kFlushedAndSimilar,
+
+  // This tier contains all block-based tables.
+  kAll,
+};
+
+// `MetadataCacheOptions` contains members indicating the desired caching
+// behavior for the different categories of metadata blocks.
+struct MetadataCacheOptions {
```

Comment:
Maybe it's just me, but I kind of find the new `MetadataCacheOptions` not as intuitive.
When configuring RocksDB's block cache pinning behaviors, I tend to think in the following pattern:
Given a type or types of meta blocks, what pinning behavior should it have?
The current config options seems to make me think in the other way:
For a specific behavior, e.g. 'pin_top_level_index', 'pin_partition', etc, what kind of SSTs should have this behavior?
It's very likely that I have missed some important piece, and let me know if I am wrong or misunderstood terribly.

Reply:
The steps I'd follow when configuring this option are:

1. Find the variable corresponding to the category of block you care about. For example, if you care about top level indexes into metadata partitions, you'd find the variable `top_level_index_pinning`.
2. Find the value of `PinningTier` according to which SSTs you want that category of blocks to be pinned. For example, if you want it to apply for all SSTs, then you'd pick `PinningTier::kAll`.

To me, that process feels aligned with the first pattern you described, but please let me know if I missed something.

Reply:
Thanks for the clarification. That makes sense.

Reply:
The old way only makes sense if you want some specific use case that matches a boolean option and have hope that the other boolean options don't interfere with yours in a weird way.

Assuming cache_index_and_filter_blocks, the behavior is boolean: pin or don't pin. To cover the other two dimensions (what kind of metadata blocks? what kind of SST files?) without excessive complexity, we make that boolean implicit.